include/dali/core/tensor_shape.h

```diff
@@ -637,6 +638,44 @@ struct TensorListShapeBase {
     shapes.resize(num_samples * sample_dim);
   }
 
+
+  ///@{
+  /**
+   * Append this TensorListShape with shapes from other TensorListShapes.
+   */
+  void append(span<const TensorListShape<sample_ndim>> tlss) {
+    int new_samples = 0;
+    for (const auto & tls : tlss) {
+      new_samples += tls.num_samples();
+    }
+    if (new_samples == 0) return;
+    auto reference_dim_value = empty() ? tlss[0].sample_dim() : sample_dim();
+    for (const auto &tls : tlss) {
+      if (tls.sample_dim() != reference_dim_value) {
+        throw std::invalid_argument(
+                "All samples in TensorListShape must have the same number of dimensions");
+      }
+    }
+    nsamples += new_samples;
+    if (sample_dim() <= 0) {
```

Comment:
```suggestion
    if (sample_dim() < 0) {
```
0 is a valid number of dimensions.

Reply:
If TLS is uninitialized and created with DynamicDimensions, it has `sample_dim==0`. On the other hand, we want code to fall into this if-clause, whenever TLS has DynamicDimensions and wasn't initialized yet. `if(empty())` includes the case, when TLS has DynamicDimensions, was initialized, but is empty, right?

Reply:
I would still consider that it's potentially dangerous to set the dynamic TLS to sample_dim = 5, leave it empty and append to it a dynamic TLS of sample_dim = 3.
This would be a hard error if those list were not empty.
That's why I preferred the `join` approach, we can probably cover both in this PR.

Reply:
It's undoable. Empty list of scalars is undistinguishable from an uninitialized list. We could make it a special case, but I think that would be a bit too much. Putting invalid sample_dim as initial value is not good either, as `resize(num_samples)` would fail.

Reply:
I think it's better to have it work as consistent as possible.